File: src/main/java/com/patrickanker/chatterbox/PermissionsManager.java
package com.patrickanker.chatterbox;

import com.patrickanker.chatterbox.api.Messenger;
import com.patrickanker.chatterbox.core.messengers.ConsoleMessenger;
import com.patrickanker.chatterbox.core.messengers.PlayerMessenger;
import com.patrickanker.chatterbox.util.OptionalConsumer;
import net.milkbowl.vault.permission.Permission;
import org.bukkit.Bukkit;

import java.util.Optional;

public final class PermissionsManager {
    private static final String ALL_WORLDS = "*";

    private Permission permissionHandler = null;
    private boolean servicesChecked      = false;

    public PermissionsManager() {
        getHandler().ifPresent(h -> Chatterbox.logInfo("Captured Vault permissions"))
                .ifNotPresent(() -> Chatterbox.logInfo("Defaulting to Bukkit \"superperms\""));
    }

    private OptionalConsumer<Permission> getHandler() {
        // Prevents multiple calls to Bukkit, which may be expensive
        if (!servicesChecked) {
            permissionHandler = Bukkit.getServer().getServicesManager().getRegistration(Permission.class).getProvider();
            servicesChecked = true;
        }

        return OptionalConsumer.of(Optional.ofNullable(permissionHandler));
    }

    public boolean hasPermission(Messenger messenger, String permission) {
        return hasPermission(messenger, permission, ALL_WORLDS);
    }

    public boolean hasPermission(Messenger messenger, String permission, String worldName) {
        // Check if the messenger is from the console or from a player
        if (messenger instanceof ConsoleMessenger) {
            return true;
        }

        PlayerMessenger pMsgr = (PlayerMessenger) messenger;

        if (getHandler().isPresent()) {
            Permission handler = getHandler().get();

            if (worldName.equals(ALL_WORLDS)) {
                if (pMsgr.getPlayer().isPresent()) {
                    return handler.playerHas(pMsgr.getPlayer().get(), permission);
                } else {
                    if (pMsgr.getOfflinePlayer().isPresent()) {
                        return handler.playerHas(null, pMsgr.getOfflinePlayer().get(), permission);
                    } else {
                        return false;
                    }
                }
            } else {
                if (pMsgr.getOfflinePlayer().isPresent()) {
                    return handler.playerHas(worldName, pMsgr.getOfflinePlayer().get(), permission);
                } else {
                    return false;
                }
            }
        } else {
            // Default to Bukkit permissions. Per-world permissions not checked
            if (pMsgr.getPlayer().isPresent()) {
                return pMsgr.getPlayer().get().hasPermission(permission);
            } else {
                return false;
            }
        }
    }

    public boolean inGroup(PlayerMessenger pMsgr, String groupName) {
        if (getHandler().isPresent()) {
            if (pMsgr.getPlayer().isPresent()) {
                return getHandler().get().playerInGroup(pMsgr.getPlayer().get(), groupName);
            } else {
                // This getOfflinePlayer().get() call is acceptable since the Player is present
                return getHandler().get().playerInGroup(null, pMsgr.getOfflinePlayer().get(), groupName);
            }
        } else {
            return hasPermission(pMsgr, "group." + groupName);
        }
    }
}

File: src/main/kotlin/chatterbox/PermissionsManager.kt
package chatterbox

import chatterbox.api.Messenger
import chatterbox.core.messengers.ConsoleMessenger
import chatterbox.core.messengers.PlayerMessenger
import net.milkbowl.vault.permission.Permission
import org.bukkit.Bukkit

private var ALL_WORLDS = "*"

object PermissionsManager {

    private var servicesChecked = false

    private var permissionHandler: Permission? = null
        get() {
            if (!servicesChecked) {
                field = Bukkit.getServer().servicesManager.getRegistration(Permission::class.java).provider
                servicesChecked = true
            }

            return field
        }

    init {
        println(if (permissionHandler != null) "Captured Vault permissions" else "Defaulting to Bukkit \"Superperms\"")
    }

    fun hasPermission(messenger: Messenger, permission: String, worldName: String = ALL_WORLDS): Boolean {
        if (messenger is ConsoleMessenger) {
            return true
        }

        val pMsgr: PlayerMessenger = messenger as PlayerMessenger

        return if (permissionHandler != null) {
            val perm = permissionHandler as Permission

            if (worldName == ALL_WORLDS) {
                return if (pMsgr.onlinePlayer != null) {
                    perm.playerHas(pMsgr.onlinePlayer, permission)
                } else {
                    if (pMsgr.offlinePlayer != null) {
                        perm.playerHas(null, pMsgr.offlinePlayer, permission)
                    } else {
                        false
                    }
                }
            } else {
                return if (pMsgr.offlinePlayer != null) {
                    perm.playerHas(worldName, pMsgr.offlinePlayer, permission)
                } else {
                    false
                }
            }
        } else {
            pMsgr.onlinePlayer?.hasPermission(permission) ?: false
        }
    }

    fun inGroup(playerMessenger: PlayerMessenger, groupName: String): Boolean {
        return if (permissionHandler != null) {
            val perm = permissionHandler as Permission

            return if (playerMessenger.onlinePlayer != null) {
                perm.playerInGroup(playerMessenger.onlinePlayer, groupName)
            } else {
                perm.playerInGroup(null, playerMessenger.offlinePlayer, groupName)
            }
        } else {
            hasPermission(playerMessenger, "group.$groupName")
        }
    }
}